hudi-client/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngradeUtil.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.table.upgrade;
+
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.HoodieTableVersion;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.hadoop.fs.FileUtil;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.io.IOException;
+
+/**
+ * Helper class to assist in upgrading/downgrading Hoodie when there is a version change.
+ */
+public class UpgradeDowngradeUtil {
```

Comment:
@nsivabalan Can a user control the hoodie layout version manually from the HoodieWriteConfig. Say, choose the older timeline layout for 0.6.0 in which case there is no need to upgrade ?

Reply:
@nsivabalan Can you point me to the part of the code that you are referring to ?

Reply:
We should upgrade no matter what - regardless of marker based rollback being on/off, regardless of timeline layout version. as of now, timeline layout version is a config that can be controlled manually. in a future version, say 1->2 we can force migrate timeline line layout if need be. 

Meta point is: making the decision to upgrade/downgrade based on write config is a problematic thing IMO. It makes reasoning with actions done during upgrade be based on configs, which can be changed in subsequent writes. We need to be able to reason about state of dataset after upgrade, and convince ourselves that any config change after that point would work well. 
i.e even if markerBasedRollback=disabled during upgrade, we prepare the dataset such that it can be turned on at any given point from there on.